README
NAME
    Class::Dot - Simple way of creating accessor methods.

VERSION
    This document describes Class::Dot version v1.0.

SYNOPSIS
         package Animal::Mammal::Carnivorous::Cat;
 
         use Class::Dot qw( :std );
 
         # A cat's properties, with their default values and type of data.
         property gender      => isa_String('male');
         property memory      => isa_Hash;
         property state       => isa_Hash(instinct => 'hungry');
         property family      => isa_Array;
         property dna         => isa_Data;
         property action      => isa_Data;
         property colour      => isa_Int(0xfeedface);
         property fur         => isa_Array('short');
 
          sub new {
             my ($class, $gender) = @_;
             my $self    = { }; # Must be anonymous hash for Class::Dot to work.
             bless $self, $class;
 
             $self->set_gender($gender);
 
             warn sprintf('A new cat is born, it is a %s. Weeeeh!',
                 $self->gender
             );
 
             return $self;
         }
 
         sub run {
             while (1) {
                 die if $self->state->{dead};
             }
         }
 
         package main;
 
         my $albert = new Animal::Mammal::Carnivorous::Cat('male');
         $albert->memory->{name} = 'Albert';
         $albert->state->{appetite} = 'insane';
         $albert->set_fur([qw(short thin shiny)]);
         $albert->set_action('hunting');
 
         my $lucy = new Animal::Mammal::Carnivorous::Cat('female');
         $lucy->memory->{name} = 'Lucy';
         $lucy->state->{instinct => 'tired'};
         $lucy->set_fur([qw(fluffy long)]);
         $lucy->set_action('sleeping');
 
         push @{ $lucy->family   }, [$albert];
         push @{ $albert->family }, [$lucy  ];

DESCRIPTION
    A simple module for creating accessor methods with default types.

SUBROUTINES/METHODS
  CLASS METHODS
   "property($property, $default_value)"
    Example:

         property foo => isa_String('hello world');
 
         property bar => isa_Int(303);

    will create the methods:

         foo( )
         set_foo($value)
 
         bar( )
         set_bar($value)

    with default return values -hello world- and -303-.

   "isa_String($default_value)"
    The property is a string.

   "isa_Int($default_value)"
    The property is a number.

   "isa_Array(@default_values)"
    The property is an array.

   "isa_Hash(%default_values)"
    The property is an hash.

   "isa_Object($kind)"
    The property is a object. (Does not really set a default value.).

   "isa_Data()"
    The property is of a not yet defined data type.

   "isa_Code()"
    The property is a subroutine reference.

   "isa_File()"
  INSTANCE METHODS
   "->properties_for_class($class)"
    Return the list of properties for a class/object that uses the powers.

  PRIVATE CLASS METHODS
   "_create_get_accessor($property, $default_value)"
    Create the set accessor for a property. Returns a code reference to the
    new setter method. It has to be installed into the callers package
    afterwards.

   "_create_set_accessor($property)"
    Create the get accessor for a property. Returns a code reference to the
    new getter method. It has to be installed into the callers package
    afterwards.

DIAGNOSTICS
  * You tried to set a value with "foo()". Did you mean "set_foo()"
    Self-explained?

CONFIGURATION AND ENVIRONMENT
    This module requires no configuration file or environment variables.

DEPENDENCIES
    *   version

INCOMPATIBILITIES
    None known.

BUGS AND LIMITATIONS
    No bugs have been reported.

    Please report any bugs or feature requests to bug-class-dot@rt.cpan.org,
    or through the web interface at CPAN Bug tracker.

SEE ALSO
  Class::InsideOut
AUTHOR
    Ask Solem, ask@0x61736b.net.

LICENSE AND COPYRIGHT
    Copyright (c), 2007 Ask Solem ask@0x61736b.net.

    All rights reserved.

    This library is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself, either Perl version 5.8.6 or, at
    your option, any later version of Perl 5 you may have available.

DISCLAIMER OF WARRANTY
    BECAUSE THIS SOFTWARE IS LICENSED FREE OF CHARGE, THERE IS NO WARRANTY
    FOR THE SOFTWARE, TO THE EXTENT PERMITTED BY APPLICABLE LAW. EXCEPT WHEN
    OTHERWISE STATED IN WRITING THE COPYRIGHT HOLDERS AND/OR OTHER PARTIES
    PROVIDE THE SOFTWARE "AS IS" WITHOUT WARRANTY OF ANY KIND, EITHER
    EXPRESSED OR IMPLIED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
    WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE. THE
    ENTIRE RISK AS TO THE QUALITY AND PERFORMANCE OF THE SOFTWARE IS WITH
    YOU. SHOULD THE SOFTWARE PROVE DEFECTIVE, YOU ASSUME THE COST OF ALL
    NECESSARY SERVICING, REPAIR, OR CORRECTION.

    IN NO EVENT UNLESS REQUIRED BY APPLICABLE LAW OR AGREED TO IN WRITING
    WILL ANY COPYRIGHT HOLDER, OR ANY OTHER PARTY WHO MAY MODIFY AND/OR
    REDISTRIBUTE THE SOFTWARE AS PERMITTED BY THE ABOVE LICENSE, BE LIABLE
    TO YOU FOR DAMAGES, INCLUDING ANY GENERAL, SPECIAL, INCIDENTAL, OR
    CONSEQUENTIAL DAMAGES ARISING OUT OF THE USE OR INABILITY TO USE THE
    SOFTWARE (INCLUDING BUT NOT LIMITED TO LOSS OF DATA OR DATA BEING
    RENDERED INACCURATE OR LOSSES SUSTAINED BY YOU OR THIRD PARTIES OR A
    FAILURE OF THE SOFTWARE TO OPERATE WITH ANY OTHER SOFTWARE), EVEN IF
    SUCH HOLDER OR OTHER PARTY HAS BEEN ADVISED OF THE POSSIBILITY OF SUCH
    DAMAGES.

    # Local Variables: # mode: cperl # cperl-indent-level: 4 # fill-column:
    78 # End: # vim: expandtab tabstop=4 shiftwidth=4 shiftround



Changes
Revision history for Class-Dot

2.0_15 Tue Feb 05 17:15:00 GMT+1 2007 [asksh@cpan.org]

    - Fixed bug with the Object type constraint where the default's
      value's instance would be constructed with ->new(q{}) instead of
      ->new(()). Patch thanks to Cosimo Streppone [cosimo@opera.com].

    - dotperltags now supports the has() keyword.

    - Fixed bugs with the Hash type constraint.

2.0_12 Mon Jan 07 21:30:00 GMT+1 2007 [asksh@cpan.org]

    - The new() constructor now supports other argument types than a hash
      reference.

    - Dependencies now reflecting the state of Debian stable.

2.0_11 Mon Jan 07 18:39:00 GMT+1 2007 [asksh@cpan.org]

    - Fixed a bug in the Object types default value factory where the
      arguments sometimes was evaluated the wrong way (sometimes they
      were suddently evaluated in scalar context instead of list context).

2.0_10 Mon Jan 07 13:17:00 GMT+1 2007 [asksh@cpan.org]

    - No API Changes.

    - Added policies: Class::Dot::Policy::Mutator,
      Class::Dot::Policy::Constrained and Class::Dot::Policy::Affordance.

    - Attributes now can have choose between different privacy settings:
        public (alias: rw), private (alias: ro) and writeonly (alias: wo).
      This means that a public attribute will have both set and get accessor,
      a private attribute will only have get accessor, and a writeonly
      accessor will only have a set accessor.

    - Lots of new tests. Tests has been categorized in a additional directory
      level. Old 1.x tests has been moved to t/010-std_1.x_tests/, New 2.x
      tests is now in t/020-std_2.x_tests/. Also some tests from Moose has
      been adopted (thanks to the Moose team!).

    - More documentation. The documentation is still not complete though.

    - In older perls (atleast in 5.6.2) any Class->isa('UNIVERSAL') returns
      true, so had to fix Class Name to use the way Class::Plugin::Util uses instead.

    - Changed dependency Test::Exception from recommended to build_requires.

    - Registry variables now lexical instead of global.

    - Massive clean-up. The huge Class::Dot module now split up into a vast
      number of modules, making Class::Dot a lot more extensible.

    - All Class::Dot classes now inherits from Class::Dot::Object.

    - All Class::Dot classes now has their own metaclass instance.
  
    - The base class for all types (DotX) now renamed to Class::Dot::Type.
     
    - Class::Dot::Types now renamed to Class::Dot::Typemap.

    - Typemap generator now does type inheritance lookup to find fields
      (default_value & constraint).

    - Mutators will be created if neither -setter_prefix or -getter_prefix is
      set. (See Class::Dot::Policy::Mutator).

    - __getattr__ now initialize the default type if the attribute has not
       been accessed by the get accessor yet.

    - Base class for types (DotX) renamed to Class::Dot::Type. The name was
      not satisfactory, sorry for any inconvenience this name has caused.
      We're now a good CPAN citizen again :)

    - All Class::Dot classes now inherits from Class::Dot::Object.
      (this might change to mixin in the future, we'll see how it works out).

2.0_09 Fri Dec 21 20:11:52 GMT+1 2007 [asksh@cpan.org]

    - lib/Class/Dot/Types.pm is now auto generated from the Typemap file.

    - *EXPERIMENTAL* Types can now have constraints. That means that you can have the set_
      accessor check if the given value actually conforms to the type of the
      property.

      Since this feature is experimental and can break code written for
      earlier versions of Class::Dot, you have to explicitly turn it on for a
      class via the use-line with the -constraints option, like this:

        use Class::Dot qw(:new -constraints);

      All Class::Dot options are on a class by class basis, so setting this
      option in one class, won't affect the others.
    
    - You can now use Class::Dot as a drop-in replacement for Moose in some
      situations.  (Depending on how much of Moose's feature-set you use).
        - Supports the Moose (has 'name' => ...) keyword.
        - Support for all Moose types.
        - Supports the Moose extends() keyword.
        - Supports type constraints via the -constraints class option.

        - Does not support roles.
        - Does not support the __PACKACKE__->meta method.
        - Does not suppoer after, augment, before etc.

    - NOTE: Now dies if you supply unknown class options on the use-line.

2.0_08 Thu Dec 20 12:44:43 GMT+1 2007 [asksh@cpan.org]

    - There was a bug in the ISA traversal. It would not fetch recursive
      inheritance. This is now fixed.

    - Class::Dot now uses mro::get_linear_isa if available. (mro was
      introduced in perl 5.9.5 and is included with the recent 5.10.0 release)

      mro::get_linear_isa is _much_ faster than the regular isa traversal
      implementation we use that is pure perl, so you might get better
      performance by upgrading to perl 5.10.0.
   
      NOTE FOR USERS OF PERL < 5.10.0 AND Class::C3

      If are already using the C3 method resolution order for your classes,
      you can install the MRO::Compat and Class::C3::XS modules to get
      a ISA traversal method implemented in C for older perl's.

      MRO::Compat's ISA traversal method for the DFS method resolution
      order is actually slower than Class::Dot's, but because of how the way
      MRO::Compat works we can't disable it if it is in installed.

    - Class::Dot now tries to load Class::Dot::XS which is a speed-up module
      in the making. It will be uploaded to CPAN when it's finished.

2.0_08 Wed Dec 19 13:49:14 GMT+1 2007 [asksh@cpan.org]
    - Reported by: srezic@cpan.org Subject: FAIL Class-Dot-2.0.0_07 MSWin32-x86-multi-thread 5.1
    Fixlog:
    Fixed problem on Win32 where extra zeros where padded to the version numbers:

    t/properties..............
    %   Failed test 'default class version is 1'

examples/Cat.pm
# $Id: Cat.pm 50 2007-11-03 21:59:03Z asksol $
# $Source$
# $Author: asksol $
# $HeadURL: https://class-dot.googlecode.com/svn/trunk/examples/Cat.pm $
# $Revision: 50 $
# $Date: 2007-11-03 22:59:03 +0100 (Sat, 03 Nov 2007) $
package Animal::Mammal::Carnivorous::Cat;

use Class::Dot qw( :std );

# A cat's properties, with their default values and type of data.
property gender => isa_String('male');
property memory => isa_Hash;
property state  => isa_Hash( instinct => 'hungry' );
property family => isa_Array;
property dna    => isa_Data;
property action => isa_Data;
property colour => isa_Int(0xfeedface);
property fur    => isa_Array('short');

sub new {
    my ( $class, $gender ) = @_;
    my $self = {};    # Must be anonymous hash for Class::Dot to work.
    bless $self, $class;

    $self->set_gender($gender);

    warn sprintf(
        'A new cat is born, it is a %s. Weeeeh!',
        $self->gender
    );

    return $self;
}

sub run {
    while (1) {
        die if $self->state->{dead};
    }
}

package main;

my $albert = new Animal::Mammal::Carnivorous::Cat('male');
$albert->memory->{name}    = 'Albert';
$albert->state->{appetite} = 'insane';
$albert->set_fur( [qw(short thin shiny)] );
$albert->set_action('hunting');

my $lucy = new Animal::Mammal::Carnivorous::Cat('female');
$lucy->memory->{name} = 'Lucy';
$lucy->state->{ instinct => 'tired' };
$lucy->set_fur( [qw(fluffy long)] );
$lucy->set_action('sleeping');

push @{ $lucy->family },   [$albert];
push @{ $albert->family }, [$lucy];

1;

# Local Variables:
#   mode: cperl
#   cperl-indent-level: 4
#   fill-column: 78
# End:
# vim: expandtab tabstop=4 shiftwidth=4 shiftround
